Index out of range when changing board size

**Describe the bug**
When I update the size of the board from 10x10 to 7x6, the game crashes after a move.

**To Reproduce**
Steps to reproduce the behavior:
1. Go to `./cmd/four-in-a-row/main.go`
2. Change game board to `g := game.NewGame(7, 6)`.
3. Run game
4. See error

**Error Message**
```
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/gokultp/go-four-in-a-row/pkg/game.(*Game).isWon(0xc0000f4000, 0x4, 0x5, 0x1, 0x10c7cac)
	/Users/berto/go/src/github.com/gokultp/go-four-in-a-row/pkg/game/game.go:132 +0xa0a
github.com/gokultp/go-four-in-a-row/pkg/game.(*Game).addEntry(0xc0000f4000, 0x4, 0x1)
	/Users/berto/go/src/github.com/gokultp/go-four-in-a-row/pkg/game/game.go:123 +0x15b
github.com/gokultp/go-four-in-a-row/pkg/game.(*Game).Input(0xc0000f4000, 0x4)
	/Users/berto/go/src/github.com/gokultp/go-four-in-a-row/pkg/game/game.go:245 +0x3e
main.main()
	/Users/berto/go/src/github.com/gokultp/go-four-in-a-row/cmd/four-in-a-row/main.go:38 +0x1bc
exit status 2
```

**Additional context**
I'm pretty sure it has to do with the [vertical check using width](https://github.com/gokultp/go-four-in-a-row/blob/master/pkg/game/game.go#L132). I changed it to use `g.Height` and it seemed fixed. There were other issues not detecting diagonal wins after that. It's inconsistent and tedious to replicate.

--- a/.gitignore
+++ b/.gitignore
@@ -1,3 +1,4 @@
 vendor/
 four-in-a-row
 .idea
+dump.txt

--- a/cmd/four-in-a-row/main.go
+++ b/cmd/four-in-a-row/main.go
@@ -2,6 +2,7 @@ package main
 
 import (
 	"flag"
+	"os"
 
 	"github.com/gokultp/go-four-in-a-row/pkg/ai"
 	"github.com/gokultp/go-four-in-a-row/pkg/game"
@@ -39,6 +40,17 @@ func main() {
 	m := game.NewManager()
 	g := game.NewGame(width, height, 0, 0)
 	g.SplashScreen()
+
+	defer func() {
+		if r := recover(); r != nil {
+			f, err := os.Create("dump.txt")
+			if err == nil {
+				g.WriteState(f)
+				f.Close()
+			}
+			panic(r)
+		}
+	}()
 
 loop:
 	for {

--- a/pkg/game/game.go
+++ b/pkg/game/game.go
@@ -1,7 +1,9 @@
 package game
 
 import (
+	"bytes"
 	"context"
+	"fmt"
 	"time"
 
 	termbox "github.com/nsf/termbox-go"
@@ -169,7 +171,7 @@ func (g *Game) isWon(col, row, player int) bool {
 	// check vertically
 	wonState := make([][]int, 0)
 	count := 0
-	for i := row; i < g.Width && g.State[i][col] == player; i++ {
+	for i := row; i < g.Height && g.State[i][col] == player; i++ {
 		count++
 		wonState = append(wonState, []int{i, col})
 	}
@@ -182,7 +184,7 @@ func (g *Game) isWon(col, row, player int) bool {
 	wonState = make([][]int, 0)
 	count = 0
 	i := col
-	for ; i > 0 && g.State[row][i] == player; i-- {
+	for ; i >= 0 && g.State[row][i] == player; i-- {
 	}
 	i++
 	// check horizontally right
@@ -200,7 +202,7 @@ func (g *Game) isWon(col, row, player int) bool {
 	wonState = make([][]int, 0)
 	i = 0
 	count = 0
-	for ; col-i > 0 && row-i > 0 && g.State[row-i][col-i] == player; i++ {
+	for ; col-i >= 0 && row-i >= 0 && g.State[row-i][col-i] == player; i++ {
 	}
 	i--
 
@@ -218,10 +220,10 @@ func (g *Game) isWon(col, row, player int) bool {
 	wonState = make([][]int, 0)
 	i = 0
 	count = 0
-	for ; col+i < g.Width && row-i > 0 && g.State[row-i][col+i] == player; i++ {
+	for ; col+i <= g.Width && row-i >= 0 && g.State[row-i][col+i] == player; i++ {
 	}
 	i--
-	for ; col+i > 0 && row-i < g.Height && g.State[row-i][col+i] == player; i-- {
+	for ; col+i >= 0 && row-i <= g.Height && g.State[row-i][col+i] == player; i-- {
 		wonState = append(wonState, []int{row - i, col + i})
 		count++
 	}
@@ -330,4 +332,12 @@ func (g *Game) SplashScreen() {
 			time.Sleep(time.Millisecond * 500)
 		}
 	}(g)
+}
+
+func (g *Game) String() string {
+	var buf bytes.Buffer
+	for _, cols := range g.State {
+		fmt.Fprintf(&buf, "%v\n", cols)
+	}
+	return buf.String()
 }

--- a/pkg/game/game.go
+++ b/pkg/game/game.go
@@ -4,9 +4,11 @@ import (
 	"bytes"
 	"context"
 	"fmt"
+	"io"
+	"strconv"
 	"time"
 
-	termbox "github.com/nsf/termbox-go"
+	"github.com/nsf/termbox-go"
 )
 
 // PlayerChar defines how should each player entries rendered
@@ -223,7 +225,7 @@ func (g *Game) isWon(col, row, player int) bool {
 	for ; col+i <= g.Width && row-i >= 0 && g.State[row-i][col+i] == player; i++ {
 	}
 	i--
-	for ; col+i >= 0 && row-i <= g.Height && g.State[row-i][col+i] == player; i-- {
+	for ; col+i >= 0 && row-i < g.Height && g.State[row-i][col+i] == player; i-- {
 		wonState = append(wonState, []int{row - i, col + i})
 		count++
 	}
@@ -334,10 +336,19 @@ func (g *Game) SplashScreen() {
 	}(g)
 }
 
-func (g *Game) String() string {
+func (g *Game) StateString() string {
 	var buf bytes.Buffer
 	for _, cols := range g.State {
-		fmt.Fprintf(&buf, "%v\n", cols)
+		buf.WriteRune('`')
+		for _, c := range cols {
+			buf.WriteString(strconv.Itoa(c))
+		}
+		buf.WriteString("`,\n")
 	}
 	return buf.String()
+}
+
+func (g *Game) WriteState(w io.Writer) {
+	fmt.Fprintf(w, "+%v\n", g)
+	w.Write([]byte(g.StateString()))
 }

--- a/pkg/game/game_test.go
+++ b/pkg/game/game_test.go
@@ -36,15 +36,15 @@ func TestGame_isWon(t *testing.T) {
 		col    int
 		row    int
 		player int
-		state []string
+		state  []string
 		want   bool
 	}{
 		"5x4 game vertical 2 item": {
-			name: "",
-			game: NewGame(5, 4, 0, 0),
-			want: false,
-			col: 2,
-			row: 1,
+			name:   "",
+			game:   NewGame(5, 4, 0, 0),
+			want:   false,
+			col:    2,
+			row:    1,
 			player: 1,
 			state: []string{
 				`00000`,
@@ -54,11 +54,11 @@ func TestGame_isWon(t *testing.T) {
 			},
 		},
 		"5x4 game vertical 4 item win": {
-			name: "",
-			game: NewGame(5, 4, 0, 0),
-			want: true,
-			col: 2,
-			row: 0,
+			name:   "",
+			game:   NewGame(5, 4, 0, 0),
+			want:   true,
+			col:    2,
+			row:    0,
 			player: 1,
 			state: []string{
 				`00100`,
@@ -68,11 +68,11 @@ func TestGame_isWon(t *testing.T) {
 			},
 		},
 		"5x4 game vertical 4 item not win": {
-			name: "",
-			game: NewGame(5, 4, 0, 0),
-			want: false,
-			col: 1,
-			row: 0,
+			name:   "",
+			game:   NewGame(5, 4, 0, 0),
+			want:   false,
+			col:    1,
+			row:    0,
 			player: 1,
 			state: []string{
 				`00100`,
@@ -82,11 +82,11 @@ func TestGame_isWon(t *testing.T) {
 			},
 		},
 		"6x4 game horizontal 4 from left": {
-			name: "",
-			game: NewGame(6, 4, 0, 0),
-			want: true,
-			col: 3,
-			row: 1,
+			name:   "",
+			game:   NewGame(6, 4, 0, 0),
+			want:   true,
+			col:    3,
+			row:    1,
 			player: 1,
 			state: []string{
 				`000000`,
@@ -96,11 +96,11 @@ func TestGame_isWon(t *testing.T) {
 			},
 		},
 		"6x4 game horizontal 4 from right": {
-			name: "",
-			game: NewGame(6, 4, 0, 0),
-			want: true,
-			col: 4,
-			row: 1,
+			name:   "",
+			game:   NewGame(6, 4, 0, 0),
+			want:   true,
+			col:    4,
+			row:    1,
 			player: 1,
 			state: []string{
 				`000000`,
@@ -110,11 +110,11 @@ func TestGame_isWon(t *testing.T) {
 			},
 		},
 		"6x4 game horizontal 4 middle": {
-			name: "",
-			game: NewGame(6, 4, 0, 0),
-			want: true,
-			col: 1,
-			row: 1,
+			name:   "",
+			game:   NewGame(6, 4, 0, 0),
+			want:   true,
+			col:    1,
+			row:    1,
 			player: 1,
 			state: []string{
 				`000000`,
@@ -124,11 +124,11 @@ func TestGame_isWon(t *testing.T) {
 			},
 		},
 		"6x4 game horizontal 3 left": {
-			name: "",
-			game: NewGame(6, 4, 0, 0),
-			want: false,
-			col: 1,
-			row: 1,
+			name:   "",
+			game:   NewGame(6, 4, 0, 0),
+			want:   false,
+			col:    1,
+			row:    1,
 			player: 1,
 			state: []string{
 				`000000`,
@@ -138,11 +138,11 @@ func TestGame_isWon(t *testing.T) {
 			},
 		},
 		"6x4 game horizontal 3 right": {
-			name: "",
-			game: NewGame(6, 4, 0, 0),
-			want: false,
-			col: 1,
-			row: 1,
+			name:   "",
+			game:   NewGame(6, 4, 0, 0),
+			want:   false,
+			col:    1,
+			row:    1,
 			player: 1,
 			state: []string{
 				`000000`,
@@ -152,11 +152,11 @@ func TestGame_isWon(t *testing.T) {
 			},
 		},
 		"6x4 game horizontal 3 middle": {
-			name: "",
-			game: NewGame(6, 4, 0, 0),
-			want: false,
-			col: 1,
-			row: 1,
+			name:   "",
+			game:   NewGame(6, 4, 0, 0),
+			want:   false,
+			col:    1,
+			row:    1,
 			player: 1,
 			state: []string{
 				`000000`,
@@ -166,11 +166,11 @@ func TestGame_isWon(t *testing.T) {
 			},
 		},
 		"6x5 game negative diagonal 3 top left": {
-			name: "",
-			game: NewGame(6, 5, 0, 0),
-			want: false,
-			col: 3,
-			row: 3,
+			name:   "",
+			game:   NewGame(6, 5, 0, 0),
+			want:   false,
+			col:    3,
+			row:    3,
 			player: 1,
 			state: []string{
 				`100000`,
@@ -181,11 +181,11 @@ func TestGame_isWon(t *testing.T) {
 			},
 		},
 		"6x5 game negative diagonal 4 top left": {
-			name: "",
-			game: NewGame(6, 5, 0, 0),
-			want: true,
-			col: 3,
-			row: 3,
+			name:   "",
+			game:   NewGame(6, 5, 0, 0),
+			want:   true,
+			col:    3,
+			row:    3,
 			player: 1,
 			state: []string{
 				`100000`,
@@ -196,11 +196,11 @@ func TestGame_isWon(t *testing.T) {
 			},
 		},
 		"6x5 game negative diagonal 4 bottom right": {
-			name: "",
-			game: NewGame(6, 5, 0, 0),
-			want: true,
-			col: 5,
-			row: 4,
+			name:   "",
+			game:   NewGame(6, 5, 0, 0),
+			want:   true,
+			col:    5,
+			row:    4,
 			player: 1,
 			state: []string{
 				`000000`,
@@ -211,11 +211,11 @@ func TestGame_isWon(t *testing.T) {
 			},
 		},
 		"6x6 game negative diagonal 4 middle": {
-			name: "",
-			game: NewGame(6, 6, 0, 0),
-			want: true,
-			col: 1,
-			row: 1,
+			name:   "",
+			game:   NewGame(6, 6, 0, 0),
+			want:   true,
+			col:    1,
+			row:    1,
 			player: 1,
 			state: []string{
 				`000000`,
@@ -227,11 +227,11 @@ func TestGame_isWon(t *testing.T) {
 			},
 		},
 		"6x5 game positive diagonal 3 top right": {
-			name: "",
-			game: NewGame(6, 5, 0, 0),
-			want: false,
-			col: 5,
-			row: 0,
+			name:   "",
+			game:   NewGame(6, 5, 0, 0),
+			want:   false,
+			col:    5,
+			row:    0,
 			player: 1,
 			state: []string{
 				`000001`,
@@ -242,11 +242,11 @@ func TestGame_isWon(t *testing.T) {
 			},
 		},
 		"6x5 game positive diagonal 4 top right": {
-			name: "",
-			game: NewGame(6, 5, 0, 0),
-			want: true,
-			col: 5,
-			row: 0,
+			name:   "",
+			game:   NewGame(6, 5, 0, 0),
+			want:   true,
+			col:    5,
+			row:    0,
 			player: 1,
 			state: []string{
 				`000001`,
@@ -257,11 +257,11 @@ func TestGame_isWon(t *testing.T) {
 			},
 		},
 		"6x5 game positive diagonal 4 bottom left": {
-			name: "",
-			game: NewGame(6, 5, 0, 0),
-			want: true,
-			col: 0,
-			row: 4,
+			name:   "",
+			game:   NewGame(6, 5, 0, 0),
+			want:   true,
+			col:    0,
+			row:    4,
 			player: 1,
 			state: []string{
 				`000000`,
@@ -272,11 +272,11 @@ func TestGame_isWon(t *testing.T) {
 			},
 		},
 		"6x6 game positive diagonal 4 middle": {
-			name: "",
-			game: NewGame(6, 6, 0, 0),
-			want: true,
-			col: 1,
-			row: 3,
+			name:   "",
+			game:   NewGame(6, 6, 0, 0),
+			want:   true,
+			col:    1,
+			row:    3,
 			player: 1,
 			state: []string{
 				`000010`,
@@ -286,6 +286,22 @@ func TestGame_isWon(t *testing.T) {
 				`000000`,
 			},
 		},
+		"6x6 game on corner": {
+			name:   "",
+			game:   NewGame(6, 6, 0, 0),
+			want:   false,
+			col:    2,
+			row:    5,
+			player: 1,
+			state: []string{
+				`000000`,
+				`000000`,
+				`000000`,
+				`000000`,
+				`200000`,
+				`101000`,
+			},
+		},
 	}
 	for name, tt := range tests {
 		t.Run(name, func(t *testing.T) {
@@ -293,7 +309,7 @@ func TestGame_isWon(t *testing.T) {
 
 			err := applyState(tt.game, tt.state)
 			if err != nil {
-			    t.Fatalf("failed to apply state: %v", err)
+				t.Fatalf("failed to apply state: %v", err)
 			}
 
 			if got := g.isWon(tt.col, tt.row, tt.player); got != tt.want {

--- a/pkg/game/game_test.go
+++ b/pkg/game/game_test.go
@@ -3,9 +3,10 @@ package game
 import (
 	"errors"
 	"fmt"
+	"strconv"
 	"testing"
 
-	termbox "github.com/nsf/termbox-go"
+	"github.com/nsf/termbox-go"
 	"github.com/smartystreets/goconvey/convey"
 )
 
@@ -26,4 +27,299 @@ func TestGetPlayerOneDisplayProps(t *testing.T) {
 		})
 	})
 
+}
+
+func TestGame_isWon(t *testing.T) {
+	tests := map[string]struct {
+		name   string
+		game   *Game
+		col    int
+		row    int
+		player int
+		state []string
+		want   bool
+	}{
+		"5x4 game vertical 2 item": {
+			name: "",
+			game: NewGame(5, 4, 0, 0),
+			want: false,
+			col: 2,
+			row: 1,
+			player: 1,
+			state: []string{
+				`00000`,
+				`00000`,
+				`00100`,
+				`00100`,
+			},
+		},
+		"5x4 game vertical 4 item win": {
+			name: "",
+			game: NewGame(5, 4, 0, 0),
+			want: true,
+			col: 2,
+			row: 0,
+			player: 1,
+			state: []string{
+				`00100`,
+				`00100`,
+				`00100`,
+				`00100`,
+			},
+		},
+		"5x4 game vertical 4 item not win": {
+			name: "",
+			game: NewGame(5, 4, 0, 0),
+			want: false,
+			col: 1,
+			row: 0,
+			player: 1,
+			state: []string{
+				`00100`,
+				`00100`,
+				`00100`,
+				`00100`,
+			},
+		},
+		"6x4 game horizontal 4 from left": {
+			name: "",
+			game: NewGame(6, 4, 0, 0),
+			want: true,
+			col: 3,
+			row: 1,
+			player: 1,
+			state: []string{
+				`000000`,
+				`111100`,
+				`000000`,
+				`000000`,
+			},
+		},
+		"6x4 game horizontal 4 from right": {
+			name: "",
+			game: NewGame(6, 4, 0, 0),
+			want: true,
+			col: 4,
+			row: 1,
+			player: 1,
+			state: []string{
+				`000000`,
+				`001111`,
+				`000000`,
+				`000000`,
+			},
+		},
+		"6x4 game horizontal 4 middle": {
+			name: "",
+			game: NewGame(6, 4, 0, 0),
+			want: true,
+			col: 1,
+			row: 1,
+			player: 1,
+			state: []string{
+				`000000`,
+				`011110`,
+				`000000`,
+				`000000`,
+			},
+		},
+		"6x4 game horizontal 3 left": {
+			name: "",
+			game: NewGame(6, 4, 0, 0),
+			want: false,
+			col: 1,
+			row: 1,
+			player: 1,
+			state: []string{
+				`000000`,
+				`111000`,
+				`000000`,
+				`000000`,
+			},
+		},
+		"6x4 game horizontal 3 right": {
+			name: "",
+			game: NewGame(6, 4, 0, 0),
+			want: false,
+			col: 1,
+			row: 1,
+			player: 1,
+			state: []string{
+				`000000`,
+				`000111`,
+				`000000`,
+				`000000`,
+			},
+		},
+		"6x4 game horizontal 3 middle": {
+			name: "",
+			game: NewGame(6, 4, 0, 0),
+			want: false,
+			col: 1,
+			row: 1,
+			player: 1,
+			state: []string{
+				`000000`,
+				`001110`,
+				`000000`,
+				`000000`,
+			},
+		},
+		"6x5 game negative diagonal 3 top left": {
+			name: "",
+			game: NewGame(6, 5, 0, 0),
+			want: false,
+			col: 3,
+			row: 3,
+			player: 1,
+			state: []string{
+				`100000`,
+				`010000`,
+				`001000`,
+				`000000`,
+				`000000`,
+			},
+		},
+		"6x5 game negative diagonal 4 top left": {
+			name: "",
+			game: NewGame(6, 5, 0, 0),
+			want: true,
+			col: 3,
+			row: 3,
+			player: 1,
+			state: []string{
+				`100000`,
+				`010000`,
+				`001000`,
+				`000100`,
+				`000000`,
+			},
+		},
+		"6x5 game negative diagonal 4 bottom right": {
+			name: "",
+			game: NewGame(6, 5, 0, 0),
+			want: true,
+			col: 5,
+			row: 4,
+			player: 1,
+			state: []string{
+				`000000`,
+				`001000`,
+				`000100`,
+				`000010`,
+				`000001`,
+			},
+		},
+		"6x6 game negative diagonal 4 middle": {
+			name: "",
+			game: NewGame(6, 6, 0, 0),
+			want: true,
+			col: 1,
+			row: 1,
+			player: 1,
+			state: []string{
+				`000000`,
+				`010000`,
+				`001000`,
+				`000100`,
+				`000010`,
+				`000000`,
+			},
+		},
+		"6x5 game positive diagonal 3 top right": {
+			name: "",
+			game: NewGame(6, 5, 0, 0),
+			want: false,
+			col: 5,
+			row: 0,
+			player: 1,
+			state: []string{
+				`000001`,
+				`000010`,
+				`000100`,
+				`000000`,
+				`000000`,
+			},
+		},
+		"6x5 game positive diagonal 4 top right": {
+			name: "",
+			game: NewGame(6, 5, 0, 0),
+			want: true,
+			col: 5,
+			row: 0,
+			player: 1,
+			state: []string{
+				`000001`,
+				`000010`,
+				`000100`,
+				`001000`,
+				`000000`,
+			},
+		},
+		"6x5 game positive diagonal 4 bottom left": {
+			name: "",
+			game: NewGame(6, 5, 0, 0),
+			want: true,
+			col: 0,
+			row: 4,
+			player: 1,
+			state: []string{
+				`000000`,
+				`000100`,
+				`001000`,
+				`010000`,
+				`100000`,
+			},
+		},
+		"6x6 game positive diagonal 4 middle": {
+			name: "",
+			game: NewGame(6, 6, 0, 0),
+			want: true,
+			col: 1,
+			row: 3,
+			player: 1,
+			state: []string{
+				`000010`,
+				`000100`,
+				`001000`,
+				`010000`,
+				`000000`,
+			},
+		},
+	}
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			g := tt.game
+
+			err := applyState(tt.game, tt.state)
+			if err != nil {
+			    t.Fatalf("failed to apply state: %v", err)
+			}
+
+			if got := g.isWon(tt.col, tt.row, tt.player); got != tt.want {
+				t.Errorf("Game.isWon() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func applyState(g *Game, state []string) error {
+	if len(state) > g.Height {
+		return fmt.Errorf("got height %d while game height is %d", len(state), g.Height)
+	}
+	for row, rowStr := range state {
+		if len(rowStr) > g.Width {
+			return fmt.Errorf("got width %d while game width is %d", len(rowStr), g.Width)
+		}
+
+		for col, char := range rowStr {
+			player, err := strconv.ParseInt(string(char), 10, 64)
+			if err != nil {
+				return fmt.Errorf("failed to parse %q at %d:%d: %v", char, row, col, err)
+			}
+			g.State[row][col] = int(player)
+		}
+	}
+
+	return nil
 }

